kolibri/content/test/test_annotation.py

```diff
@@ -0,0 +1,209 @@
+        self.assertTrue(all(
+            ContentNode.objects.exclude(kind=content_kinds.TOPIC).exclude(files=None).values_list('available', flat=True)))
+
+    def test_no_local_files_available(self):
+        LocalFile.objects.all().update(available=False)
+        with patch('kolibri.content.utils.sqlalchemybridge.get_engine', new=get_engine):
+            set_leaf_node_availability_from_local_file_availability()
+        self.assertEqual(File.objects.filter(available=True).count(), 0)
+        self.assertEqual(ContentNode.objects.exclude(kind=content_kinds.TOPIC).filter(available=True).count(), 0)
+
+    def test_one_local_file_available(self):
+        LocalFile.objects.all().update(available=False)
+        localfile = LocalFile.objects.get(id='9f9438fe6b0d42dd8e913d7d04cfb2b2')
+        localfile.available = True
+        localfile.save()
+        with patch('kolibri.content.utils.sqlalchemybridge.get_engine', new=get_engine):
+            set_leaf_node_availability_from_local_file_availability()
+        self.assertTrue(ContentNode.objects.get(id='32a941fb77c2576e8f6b294cde4c3b0c').available)
+        self.assertFalse(all(
+            ContentNode.objects.exclude(
+                kind=content_kinds.TOPIC).exclude(id='32a941fb77c2576e8f6b294cde4c3b0c').values_list('available', flat=True)))
+
+    def tearDown(self):
+        call_command('flush', interactive=False)
+        super(AnnotationFromLocalFileAvailability, self).tearDown()
+
+
+class AnnotationTreeRecursion(TransactionTestCase):
+
+    fixtures = ['content_test.json']
+
+    def test_all_content_nodes_available(self):
+        ContentNode.objects.all().update(available=False)
+        ContentNode.objects.exclude(kind=content_kinds.TOPIC).update(available=True)
+        with patch('kolibri.content.utils.sqlalchemybridge.get_engine', new=get_engine):
+            recurse_availability_up_tree()
+        self.assertTrue(all(
+            ContentNode.objects.filter(kind=content_kinds.TOPIC).values_list('available', flat=True)))
+
+    def test_no_content_nodes_available(self):
+        ContentNode.objects.exclude(kind=content_kinds.TOPIC).update(available=False)
+        with patch('kolibri.content.utils.sqlalchemybridge.get_engine', new=get_engine):
+            recurse_availability_up_tree()
+        self.assertEqual(ContentNode.objects.filter(kind=content_kinds.TOPIC).filter(available=True).count(), 0)
+
+    def test_one_content_node_available(self):
+        ContentNode.objects.all().update(available=False)
+        contentnode = ContentNode.objects.get(id='32a941fb77c2576e8f6b294cde4c3b0c')
+        contentnode.available = True
+        contentnode.save()
+        with patch('kolibri.content.utils.sqlalchemybridge.get_engine', new=get_engine):
+            recurse_availability_up_tree()
+        # Check parent is available
+        self.assertTrue(ContentNode.objects.get(id='da7ecc42e62553eebc8121242746e88a').available)
```

Comment:
Right -- was mostly just thinking for semantic readability of the test cases. But not critical in any way